drivers/serial/uart_nrfx_uarte.c

```diff
@@ -1040,69 +1119,73 @@ static int uarte_nrfx_poll_in(const struct device *dev, unsigned char *c)
  */
 static void uarte_nrfx_poll_out(const struct device *dev, unsigned char c)
 {
-	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
 	struct uarte_nrfx_data *data = get_dev_data(dev);
-	atomic_t *lock;
+	bool isr_mode = k_is_in_isr() || k_is_pre_kernel();
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+	int key;
 
-#ifdef CONFIG_UART_ASYNC_API
+	if (isr_mode) {
+		while (1) {
+			bool res;
 
-	if (data->async) {
-		while (data->async->tx_buf) {
-			/* If there is ongoing transmission, and we are in
-			 * isr, then call uarte interrupt routine, otherwise
-			 * busy wait until transmission is finished.
-			 */
-			if (k_is_in_isr()) {
+#if CONFIG_UART_ASYNC_API
+			if (data->async) {
 				uarte_nrfx_isr_async(dev);
 			}
-		}
-		/* Use tx_size as lock, this way uarte_nrfx_tx will
-		 * return -EBUSY during poll_out.
-		 */
-		lock = &data->async->tx_size;
-	} else
 #endif
-		lock = &data->poll_out_lock;
 
-	if (!k_is_in_isr()) {
-		uint8_t safety_cnt = 100;
+			uint32_t wait_us = (data->uart_config.flow_ctrl ==
+					    UART_CFG_FLOW_CTRL_NONE) ?
+					    POLL_OUT_TIMEOUT_US :
+					    POLL_OUT_FC_TIMEOUT_US;
 
-		while (atomic_cas((atomic_t *) lock,
-				(atomic_val_t) 0,
-				(atomic_val_t) 1) == false) {
-			/* k_sleep allows other threads to execute and finish
-			 * their transactions.
-			 */
-			k_msleep(1);
-			if (--safety_cnt == 0) {
+			NRFX_WAIT_FOR(is_tx_ready(dev), wait_us, 1, res);
+			if (!res) {
+				LOG_WRN("Timeout in poll_out");
+				return;
+			}
+			key = irq_lock();
+			if (is_tx_ready(dev)) {
 				break;
 			}
+
+			irq_unlock(key);
 		}
 	} else {
-		*lock = 1;
-	}
+		int safety_cnt = 1000;
 
-	/* reset transmitter ready state */
-	nrf_uarte_event_clear(uarte, NRF_UARTE_EVENT_ENDTX);
+		do {
+			/* wait arbitrary time before back off. */
+			bool res;
 
-	/* send a character */
-	nrf_uarte_tx_buffer_set(uarte, &c, 1);
-	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTTX);
+			NRFX_WAIT_FOR(is_tx_ready(dev), 100, 1, res);
 
-	/* Wait for transmitter to be ready */
-	int res;
+			if (res) {
+				key = irq_lock();
+				if (is_tx_ready(dev)) {
+					break;
+				}
 
-	NRFX_WAIT_FOR(nrf_uarte_event_check(uarte, NRF_UARTE_EVENT_ENDTX),
-		      1000, 1, res);
+				irq_unlock(key);
+			}
+			k_msleep(1);
+			if (--safety_cnt == 0) {
+				LOG_ERR("Dropping byte in poll_out");
+				return;
```

Comment:
With this still present, this PR does not really fix #26722. As stated there:
> The expected behavior for hardware flow control used with uarte_nrfx_poll_out() is that it would block forever until the other side asserts CTS.

Reply:
removed any timeout when hwfc is enable, in thread mode also removed a timeout since byte should eventually be sent (unless CTS is asserted)